text/0000-guaranteed-tco.md

```diff
@@ -0,0 +1,607 @@
+or method calls, that is supported are calls like: `become foo()`, `become foo(a)`, `become foo(a, b)`, and so on, or
+`become foo.bar()` with plain arguments. Neither the function call nor any arguments can be part of a larger expression
+such as `become foo() + 1`, `become foo(1 + 1)`, `become foo(bar())`. Additionally, there is a further restriction on
+the tail-callable functions: the function signature must exactly match that of the calling function. 
+
+Invocations of overloaded operators with at least one non-primitive argument were considered as valid targets, but were
+rejected on grounds of being too error-prone. In any case, these can still be called as methods.
+
+## Type checking
+[typechecking]: #typechecking
+A `become` statement is type-checked like a `return` statement, with the added restriction of exactly matching the
+function signatures between caller and callee. Additionally, the caller and callee **must** use the same calling
+convention.
+
+## Borrowchecking and Runtime Semantics
+[semantics]: #semantics
+A `become` expression acts as if the following events occurred in-order:
+
+1. All variables that are being passed by-value are moved to temporary storage.
+2. All local variables in the caller are destroyed according to usual Rust semantics. Destructors are called where
+   necessary. Note that values moved from in step 1 are _not_ dropped.
+3. The caller's stack frame is removed from the stack.
+4. Control is transferred to the callee's entry point.
+
+This implies that it is invalid for any references into the caller's stack frame to outlive the call. The borrow checker ensures that none of the above steps will result in the use of a value that has gone out of scope.
+
+As `become` is always in a tail position (due to being used in place of `return`), this requirement for TCO is already
+fulfilled.
+
+Example:
+```rust
+fn x() {
+    let a = Box::new(());
+    let b = Box::new(());
+    become y(a)
+}
+```
+
+Will be desugared in the following way:
+```rust
+fn x() {
+    let a = Box::new(());
+    let b = Box::new(());
+    let _tmp = a;
+    drop(b);
+    become y(_tmp)
+}
+```
+
+## Implementation
+[implementation]: #implementation
+
+A now six years old implementation for the earlier mentioned
+[RFC](https://github.com/DemiMarie/rfcs/blob/become/0000-proper-tail-calls.md) can be found at
+[DemiMarie/rust/tree/explicit-tailcalls](https://github.com/DemiMarie/rust/tree/explicit-tailcalls).
+A new implementation is planned as part of this RFC.
+
+The parser parses `become` exactly how it parses the `return` keyword. The difference in semantics is handled later.
+
+During type checking, the following are checked:
+
+1. The target of the tail call is, in fact, a simple call.
+2. The target of the tail call has the proper ABI.
+
+Later phases in the compiler assert that these requirements are met.
+
+New nodes are added in HIR and THIR to correspond to `become`. In MIR, the function call is checked that:
```

Comment:
You mean for the HIR/THIR correct? As I don't have any relevant experience here, I hope someone else can answer this question. @Amanieu do you have an opinion on this?

Reply:
I think this is compiler implementation detail invisible to the users, so it is not necessary to document it in the RFC. I personally would expect a separate node (so that, for example, optimizations can't reorder tail call out of the tail position), but this should not matter (and other compilers can choose the implementation freely).

Reply:
Yes, agreed, while the RFC template asks for ["sufficient detail"](https://github.com/phi-go/rfcs/blob/guaranteed-tco/0000-template.md#reference-level-explanation), this detail probably does not need to be specified.

Reply:
I rewrote the reference-level explanation: https://github.com/rust-lang/rfcs/pull/3407/commits/8ab2fbda329209a1c972d182d5a4462c9cd8a5f2

My goal was to only specify what is needed and leave out implementation details.